### DISCUSSION TITLE
[Suggestions] Have access to a small "database" of spells known by opposing heroes for both AI and human players

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#3444] [Ideas]
I think it would be nice if, as they meet other heroes in battle, and depending on the spells cast by opposing heroes visible on the adventure map (Dimension Door, etc.), and also previously recruited heroes who are now in another player's ranks (only the spells he knew at that time, not the ones he learned later), AI players could build up a small "database" of spells known by other heroes to elaborate their strategy of using spells and placing troops during combat (and also on the adventure map, if for example an opposing hero possesses the Dimension Door spell, before planning the movements of his heroes).
Also, with this feature, it would be possible for the human player to consult the known spells of the opposing hero during combat (observed during previous combats, etc.) in order to adapt his strategy to cast his spells and move his troops.

For example, during a combat, clicking on the portrait of the opposing hero opens his Magic Book with the spells that we have seen him use during other combats or that we know he knows because he was one of our heroes previously.

![Spells known opposing heroes 01](https://user-images.githubusercontent.com/43583994/118668829-80296880-b7f5-11eb-81d1-45aff14a7d99.png)

Related issue: https://github.com/ihhub/fheroes2/issues/3416.

**Edit 2021/11/30.**

![Heroes 2 Screen 65](https://user-images.githubusercontent.com/43583994/144089183-2eececf2-f5db-445d-ad28-bfc24cfe81ee.png)

(1) The enemy hero's magic book opened, with all the spells you have seen him use in previous combats, etc.
The information bar at the bottom of the screen shows that this is the opponent's magic book.

This image (and the text) is for explanatory purposes only.

**Comment by vincent-grosbois** [comment]
I like this idea :) but it's possible that enemy heroes only have a spell because they got it from a scroll, and so they wont have it for next battles potentially

**Reply by LeHerosInconnu** [reply]
Hello @vincent-grosbois,

> I like this idea :) but it's possible that enemy heroes only have a spell because they got it from a scroll, and so they wont have it for next battles potentially

It has to be part of the game, there can be small mistakes due to Spell Scrolls (or due to Artifacts that contain spells), ;) I'm okay with that. :)
I think it would also add some spice to the game. :D

I also thought of adding an extra bookmark to the Magic Book for the spells the player wants to use most often for each hero of the player.
So the player could have quick access to the spells the hero uses most often without having to turn many pages when the hero knows many spells.
And also an extra bookmark in the Magic Book for the spells that are used the most by each opponent hero during the combats against the player heroes and castles.

But I haven't posted it yet. :)

**Reply by zenseii** [reply]
What if your hero loses against an opposing hero and he gets your scroll? Would you mark this magic scroll spell for that hero?

**Reply by LeHerosInconnu** [reply]
@zenseii.

> What if your hero loses against an opposing hero and he gets your scroll? Would you mark this magic scroll spell for that hero?

The opposing hero captured the Spell Scroll so yes, he is supposed to know the spell. :)
Now if this hero gives this Spell Scroll to another hero, there could be a new mistake in the database. ;)
This is part of the intelligence game.

**Reply by Branikolog** [reply]
In a case enemy hero has full inventory the scroll will be lost. So another mistake in database.
If it works so unpredictable, do we really need this? :) Lets make players using their memory a little bit more for tracking this? :)

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

> In a case enemy hero has full inventory the scroll will be lost. So another mistake in database. If it works so unpredictable, do we really need this? :) Lets make players using their memory a little bit more for tracking this? :)

If there is no space left in the inventory of the opposing hero, the Spell Scroll will not be transferred and will remain in the inventory of the defeated hero. :)

**Reply by Branikolog** [reply]
A player cannot know this. :)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> A player cannot know this. :)

Oh yes, you are right, as the hero did not surrender, he is no longer in the ranks of the player. :)

> Lets make players using their memory a little bit more for tracking this? :)

On an XL map with a lot of heroes and a multi-session game over a few real days, I'm afraid the player's memory will fail. :D
In the case of the AI, this is done to improve its efficiency, and in the case of the player it is a quality of life improvement to allow him to make decisions with the information he knows (that he is supposed to know). :)

**Reply by LeHerosInconnu** [reply]
After some thought on the subject. :)

> @zenseii.
> 
> > What if your hero loses against an opposing hero and he gets your scroll? Would you mark this magic scroll spell for that hero?
> 
> The opposing hero captured the Spell Scroll so yes, he is supposed to know the spell. :)
> Now if this hero gives this Spell Scroll to another hero, there could be a new mistake in the database. ;)
> This is part of the intelligence game.

> In a case enemy hero has full inventory the scroll will be lost. So another mistake in database.

I think that it should not take into account the spells that the opposing hero could have (thanks to a Spell Scroll or thanks to an Artifact) but only the spells that we have seen him using (during a combat or on the adventure map).
